### PR TITLE
fix: Rules when associated with the "browser" env

### LINKED ISSUE #43: Leaky event listeners not flagged by 'no-leaky-event-listeners' eslint rule
Use lwc-recepies to repro: https://github.com/trailheadapps/lwc-recipes
- Upgrade to latest eslint config (package.json)
```
"@salesforce/eslint-config-lwc": "^0.6.0",
```

- Add `no-leaky-event-listeners` eslint rule(.eslintrc.json)
```
{
    "extends": ["@salesforce/eslint-config-lwc/recommended"],
    "rules": {
        "@lwc/lwc/no-leaky-event-listeners": "error"
    }
}
```

- Add a leaky event listener to one of lwc component
```
    handleTest() {}

    connectedCallback() {
        window.addEventListener('test', this.handleTest.bind(this));
    }
```

- run eslint: `npm run lint:lwc`

Expected:
Leaky event listener gets flagged as error and eslint fails

Actual:
Leaky event listener not get flagged and eslint passes

### PR DESCRIPTION
**Pull request by pmdartus**
The following rules `no-leaky-event-listeners`, `no-document-location`, and  `no-window-top` are not working as expected when the [`browser` env](https://eslint.org/docs/user-guide/configuring#specifying-environments) flag is set. All those rules expect that the `through` property on the global scope contains the list of all the unresolved referenced. This is not the case when the environment adds those identifiers to the global scope. When the `browser` environment is enabled, all the DOM APIs are part of the global scope and can't be resolved via the `through` property.

Fix #43.

### CODE REVIEW

File: lib/rules/no-document-location.js [hunk 1]
```diff
@@ -22,7 +22,13 @@ module.exports = {
     create(context) {
         return {
             'Program:exit'() {
-                const ref = context.getScope().through.find(({ identifier }) => {
+                // The top level scope is the "global" scope. Depending on the "env" property in the
+                // eslint config, references might or not be resolved at the global scope. The most
+                // reliable way to get all the unresolved references from the module is to get them
+                // from the "module" scope.
+                const moduleScope = context.getScope().childScopes[0];
```
Comment (ravijayaramappa) [inline lib/rules/no-document-location.js:29]:
```suggestion
                const moduleScope = context.getScope().childScopes.find(childScope => childScope.type === "module");
```

File: lib/rules/no-window-top.js [hunk 2]
```diff
@@ -22,7 +22,13 @@ module.exports = {
     create(context) {
         return {
             'Program:exit'() {
-                const ref = context.getScope().through.find(({ identifier }) => {
+                // The top level scope is the "global" scope. Depending on the "env" property in the
+                // eslint config, references might or not be resolved at the global scope. The most
```
Comment (ravijayaramappa) [inline lib/rules/no-window-top.js:26]:
```suggestion
                // eslint config, references might not be resolved at the global scope. The most
```

File: lib/rules/no-window-top.js [hunk 3]
```diff
@@ -22,7 +22,13 @@ module.exports = {
...
+                // reliable way to get all the unresolved references from the module is to get them
+                // from the "module" scope.
+                const moduleScope = context.getScope().childScopes[0];
```
Comment (ravijayaramappa) [inline lib/rules/no-window-top.js:29]: Same here, IMO its better to explicitly find the module scope instead of implicitly getting the first childScope

**Comment by pmdartus** [comment]
@ravijayaramappa Ready for another round of review.

File: lib/rules/no-document-location.js [hunk 4]
```diff
@@ -22,7 +22,15 @@ module.exports = {
...
             'Program:exit'() {
-                const ref = context.getScope().through.find(({ identifier }) => {
+                // The top level scope is the "global" scope. Depending on the "env" property in the
+                // eslint config, references might not be resolved at the global scope. The most
+                // reliable way to get all the unresolved references from the module is to get them
+                // from the "module" scope.
+                const moduleScope = context
+                    .getScope()
+                    .childScopes.find((childScope) => childScope.type === 'module');
+
```
Comment (jdalton) [inline lib/rules/no-document-location.js:32]: Does this work regardless of `env` property ☝️ ?

Comment (pmdartus) [reply lib/rules/no-document-location.js:32]: It should work regardless on the environment. The environment sets the variable in the `root` scope.
The `module` should contain all the unresolved variables.

**Pull request merged**